### PR TITLE
Infer masking vector sizes from the IR 

### PR DESCRIPTION
**Pull request by dcaballe**
This patch adds logic to infer the tile sizes for vectorization from the IR. It should cover most but not all of the cases. When the vector size inference fails, we fall back to the lowering_config approach. This should be NFC for generic op vectorization but it will have an impact when vectorizing convolutions and matmuls fused with other ops.

benchmarks: x86_64, comp-stats

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp [hunk 1]
```diff
@@ -164,10 +166,129 @@ static SmallVector<int64_t> getCanonicalVectorShape(func::FuncOp funcOp) {
...
+          break;
+        }
+        currentOp = subviewOp.getDynamicSize(operandDim).getDefiningOp();
+        continue;
+      }
+      if (auto affineApplyOp = dyn_cast<AffineApplyOp>(currentOp)) {
+        AffineValueMap valueMap = affineApplyOp.getAffineValueMap();
+        if (valueMap.getNumDims() != 1 || valueMap.getNumSymbols() != 1)
+          return failure();
+        currentOp = valueMap.getOperand(1).getDefiningOp();
```
Comment (vmurali) [inline compiler/src/iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp]: Why are you restricting it to 1 dimension and 1 symbol?

Comment (vmurali) [reply compiler/src/iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp]: And I am not sure what `getOperand(1)` does for an `AffineValueMap`

File: compiler/src/iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp [hunk 2]
```diff
@@ -164,10 +166,129 @@ static SmallVector<int64_t> getCanonicalVectorShape(func::FuncOp funcOp) {
...
+      continue;
+    }
+
+    auto affineMin = dyn_cast_or_null<AffineMinOp>(currentOp);
+    if (!affineMin) {
+      return failure();
+    }
+
+    auto minResults = affineMin.getAffineMap().getResults();
+    if (minResults.size() != 2) {
```
Comment (vmurali) [inline compiler/src/iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp]: I thought `AffineMin` always returns one value

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
Could we discuss this a bit more. I see this as a chicken and egg problem. You have to determine the vector size you want to use and do tiling, etc. to get to that size. Here you are inferring based off the IR resulting from that decision.
I see two downsides here
1) This is not addressing the real issue of how do you decide the vector size?
2) It is too hard-coded to the current form of the IR that is generated.... There is no requirement for the exact `affine.min`/`affine.apply`s etc. that are being looked for here to be generated by the tiling code....

File: compiler/src/iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp (hunk 1 above)
Comment (dcaballe) [reply compiler/src/iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp]: I'll add a comment that should answer your question

File: compiler/src/iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp (hunk 2 above)
Comment (dcaballe) [reply compiler/src/iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp]: Wrong var name. These are the results of the affine map.

**Comment by dcaballe** [comment]
Aarg... I replied to your concerns with an example, etc., Mahesh, and then lost the reply somehow... (sigh) Let me provide a short answer and follow-up over chat...

> Could we discuss this a bit more. I see this as a chicken and egg problem. 

Not sure I follow the chicken-and-egg characterization. 

> You have to determine the vector size you want to use and do tiling, etc. to get to that size. Here you are inferring based off the IR resulting from that decision.

When we apply tiling, we are materializing the lowering_config information in the IR. This approach is basically retrieving that information from the IR instead of looking at lowering config again. This lets us vectorizes many more cases as we don't have to worry about the lowering_config being still around when we are applying vectorization, the lowering config being duplicated in the dispatch, mapping the lowering config to other non-root operations with the same or different dimension as the root op, etc.

> This is not addressing the real issue of how do you decide the vector size?

Yes, that is not the goal of this patch. Tile size selection is a different problem. We are addressing the problem of looking at lowering config when we have already started to transform the code and the fact that we have to map the lowering config to non-root ops fused within the same dispatch.

> It is too hard-coded to the current form of the IR that is generated.... There is no requirement for the exact affine.min/affine.applys etc. that are being looked for here to be generated by the tiling code....

The approach is mostly taking advantage of the structure of the code, which is a common practice in Linalg/Structured Code Generation. We follow the U-D chain from the operand with the dynamic dim, go to its tensor.extract_slice/memref.subview, try to get the size information for that dynamic dim, look at the target size symbol (not any other), if no static, and follow through affine applies and affine.mins that define that size symbols. The affine.applies/mins are restricted to the specific flavors that feed the size symbols of tensor.extract_slice/memref.subview, and will bail out if something unexpected is found. I have run this with asserts in CI and haven't found a case that falls out of these specific cases, except for those dynamic dims that are not defined by an affine.min.

In general, I don't see a major concern here and I expect to replace this with some affine analysis in the short term. I'm talking to @matthias-springer about that.

Happy to discuss this further. Just let me know.

**Comment by dcaballe** [comment]
Let me make it more restrictive and remove the affine.apply case, which only happens for peeled loops with static shapes.

**Comment by dcaballe** [comment]
I talked to @matthias-springer and he is working on a proposal for more powerful upper-bound detection. In the meantime, I played with `linalg::getConstantUpperBoundForIndex` but it doesn't work for my case because the starting point is not and index (dimension) but a symbol (dimension size) and we have to go inevitably analyze the `tensor.extract_slice`/`memref.subviews` ops. Looking at the code it's also looking at specific affine.mins/affine.applies so I don't see a big difference with my utility other than the use of affine constraints. I could implement a `linalg::getConstantUpperBoundForDimSize` following the same approach but I think it's not worth the investment if @matthias-springer is working on something more robust and powerful. Are there any major concerns on moving forward with this after removing affine.map from the utility and adding extra checks?

**Comment by MaheshRavishankar** [comment]
Chatted with Diego offline. This might be a transient state and will move to use upstream utilities. This kind of lifting can become a bottleneck for evolution, but its OK as a transient state....
I dont have blocking concerns. Ill let @vmurali or @hanhanW provide approval...

**Review by vmurali — APPROVED** [review]
LGTM

**Comment by hanhanW** [comment]
> Chatted with Diego offline. This might be a transient state and will move to use upstream utilities. This kind of lifting can become a bottleneck for evolution, but its OK as a transient state.... I dont have blocking concerns. Ill let @vmurali or @hanhanW provide approval...

I reviewed the code, and I don't have big concerns. I've hit issues about getting static informations from affine maps, so I'm fine with implementing them in a light way as a transient state. The code overall looks good to me!

**Review by hanhanW — APPROVED** [review]

**Pull request closed without merging**